Wikipedia/Code/WMFTaskGroup.m

```diff
@@ -0,0 +1,50 @@
+#import "WMFTaskGroup.h"
+
+@interface WMFTaskGroup ()
+    
+@property (nonatomic) dispatch_group_t group;
+@property (nonatomic) NSInteger count;
+
+@end
+
+@implementation WMFTaskGroup
+
+- (instancetype)init {
+    self = [super init];
+    if (self) {
+        self.group = dispatch_group_create();
+    }
+    return self;
+}
+- (void)enter {
+    @synchronized (self) {
+        self.count++;
+        dispatch_group_enter(_group);
+    }
+}
+
+- (void)leave {
+    @synchronized (self) {
+        if (self.count > 0) {
+            self.count--;
+            dispatch_group_leave(self.group);
+        } else {
+            DDLogError(@"Mismatched leave for group: %@", self);
+        }
+    }
+}
+
+- (void)waitInBackgroundWithTimeout:(NSTimeInterval)timeout completion:(nonnull dispatch_block_t)completion {
+    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_BACKGROUND, 0), ^{
```

Comment:
@coreyfloyd @montehurd correct - the synchronous wait is only for having a timeout.  `dispatch_group_notify` is used when there is no timeout. The timeout method isn't used in the codebase, I'd be OK with removing it. I added it so we could consider it, but I think there's more risk in it timing out early based on an arbitrary guess at timeout length. The real danger is a crash due to over-leaving the dispatch group.